File: Ch4/01.pwd/spwd.rs
use std::fs::{metadata, Metadata,read_dir};
use std::os::unix::fs::MetadataExt;
use std::process;
use std::env::set_current_dir;

fn get_inode(filename: &str) -> u64{
    let md: Metadata = match metadata(filename) {
        Ok(md) => md,
        Err(msg) => {
            eprintln!("cannot stat {} {:?}", filename, msg);
            process::exit(-1);
        }
    };
    md.ino()
}

fn inode_to_name(this_inode: u64) -> String {
    if let Ok(rd) = read_dir(".") {
        for item in rd {
            if let Ok(de) = item {
                match de.metadata() {
                    Ok(meta) => {
                        if meta.ino() == this_inode {
                            return de.file_name().into_string().expect("cannot convert this OsStirng to String");
                        }
                    },
                    Err(msg) => {
                        eprintln!("cannot inspect some entries in current dir {:?}", msg);
                    }
                }
            }else{
                eprintln!("cannot inspect some entries in current dir");
            }
        }
    }
    eprintln!("error looking for inode:{}", this_inode);
    process::exit(1);
}

fn print_path_to(inode: u64) {
    let mut my_inode: u64 = 0;
    let mut name: String = String::new();

    if get_inode("..") != inode {
        if set_current_dir("..").is_err() {
            eprintln!("cannot change working directory to the parent directory");
            process::exit(-1);
        }
        name = inode_to_name(inode);
        my_inode = get_inode(".");
        print_path_to(my_inode);
        print!("/{}", name);
    }
}

fn main() {
    print_path_to(get_inode("."));
    println!();
}

File: Ch4/01.pwd/spwd.c
/*
 * spwd.c: a simplified version of pwd
 * 
 * starts in current directory and recursively
 * climbs up to root of filesystem, prints top part
 * then prints currnet part
 * 
 * uses readdir() to get info about each thing
*/

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

ino_t get_inode(char *);
void print_path_to(ino_t);
void inode_to_name(ino_t, char *, int);

int main(){
    print_path_to(get_inode("."));
    putchar('\n');
    return 0;
}

/*
 * returns inode number of the file
*/
ino_t get_inode(char * filename) {
    struct stat buf;
    if (stat(filename, &buf) == -1) {
        fprintf(stderr, "cannot stat ");
        perror(NULL);
        exit(-1);
    }
    return buf.st_ino;
}

/*
 * looks through current directory for a file with this inode
 * number and copied its name into namebuf
*/
void inode_to_name(ino_t this_inode, char * namebuf, int buflen) {
    DIR * dir_ptr = NULL;
    struct dirent * dirent_ptr = NULL;

    if ((dir_ptr = opendir(".")) == NULL) {
       perror(NULL);
       exit(-1);
    }

    while ((dirent_ptr=readdir(dir_ptr)) != NULL) {
        struct stat buf;
        stat(dirent_ptr->d_name, &buf);
        if (buf.st_ino == this_inode) {
            strncpy(namebuf, dirent_ptr->d_name, buflen);
            closedir(dir_ptr);
            namebuf[buflen-1] = '\0';
            return;
        }
    }
    fprintf(stderr, "error looking for inode: %ld\n", this_inode);
    exit(1);
}

/*
 * prints each leading down to an object with this inode
 * kindof recursive
*/
void print_path_to(ino_t inode) {
    ino_t my_inode = 0;
    char its_name[256];

    if (get_inode("..") != inode) {
        chdir("..");
        inode_to_name(inode, its_name, 256);
        my_inode = get_inode(".");
        print_path_to(my_inode);
        printf("/%s", its_name);
    }
}